opendir.c

```diff
@@ -47,23 +47,91 @@ struct watchman_dir_handle {
   struct watchman_dir_ent ent;
 };
 
+/* Opens a directory, strictly prohibiting opening any symlinks
+ * in any component of the path on systems that make this
+ * reasonable and easy to do.
+ * We do this by opening / and then walking down the tree one
+ * directory component at a time.
+ */
+static int open_dir_as_fd_no_symlinks(const char *path, int flags) {
+#ifdef HAVE_OPENAT
+  char *pathcopy;
+  int fd = -1;
+  int err;
+  char *elem, *next;
+
+  if (path[0] != '/') {
+    errno = EINVAL;
+    return -1;
+  }
+
+  pathcopy = strdup(path);
+  if (!pathcopy) {
+    return -1;
+  }
+
+  fd = open("/", O_NOFOLLOW | O_CLOEXEC);
```

Comment:
Yeah, it should be unpossible.  I don't think it hurts to add this check here and helps underscore the intent